src/python/pants/backend/native/targets/native_artifact.py

```diff
@@ -1,14 +1,20 @@
 # Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+from dataclasses import dataclass
 from hashlib import sha1
 
 from pants.base.payload_field import PayloadField
-from pants.util.objects import datatype
 
 
-class NativeArtifact(datatype(['lib_name']), PayloadField):
+# NB: We do not decorate this with @frozen_after_init because PayloadField must still mutate
```

Comment:
Are we sure the benefits of dataclass outweigh the cons in this particular case? Anyway we can be a bit more manual with this type and remove this gotcha?

Reply:
I agree this isn't great. Modifying it so that we still use `dataclass`, but implement our own custom `__hash__()` that does the same thing as `unsafe_hash=True` but does it in a much more explicit way. Also overriding `__setattr__()` to ensure that we never accidentally change the hash.